examples/vision/deit.py

```diff
@@ -0,0 +1,575 @@
+import tensorflow_addons as tfa
+import tensorflow_datasets as tfds
+import tensorflow_hub as hub
+from tensorflow import keras
+from tensorflow.keras import layers
+
+tfds.disable_progress_bar()
+tf.keras.utils.set_random_seed(42)
+
+"""
+## Constants
+"""
+
+# Model
+MODEL_TYPE = "deit_distilled_tiny_patch16_224"
+RESOLUTION = 224
+PATCH_SIZE = 16
+NUM_PATCHES = (RESOLUTION // PATCH_SIZE) ** 2
+LAYER_NORM_EPS = 1e-6
+PROJECTION_DIM = 192
+NUM_HEADS = 3
+NUM_LAYERS = 12
+MLP_UNITS = [
+    PROJECTION_DIM * 4,
+    PROJECTION_DIM,
+]
+DROPOUT_RATE = 0.0
+DROP_PATH_RATE = 0.1
+
+# Training
+NUM_EPOCHS = 20
+BASE_LR = 0.0005
+WEIGHT_DECAY = 0.0001
+
+# Data
+BATCH_SIZE = 256
+AUTO = tf.data.AUTOTUNE
+NB_CLASSES = 5
+
+"""
+## Load the `tf_flowers` dataset and prepare preprocessing utilities
+
+The authors use an array of different augmentation techniques including MixUp
+([Zhang et al.](https://arxiv.org/abs/1710.09412)),
+RandAugment ([Cubuk et al.](https://arxiv.org/abs/1909.13719)),
+and so on. However, to keep the example simple to work through, we'll discard them.
+"""
+
+
+def preprocess_dataset(is_training=True):
+    def _pp(image, label):
+        if is_training:
+            # Resize to a bigger spatial resolution and take the random
+            # crops.
+            image = tf.image.resize(image, (RESOLUTION + 20, RESOLUTION + 20))
+            image = tf.image.random_crop(image, (RESOLUTION, RESOLUTION, 3))
+            image = tf.image.random_flip_left_right(image)
+        else:
+            image = tf.image.resize(image, (RESOLUTION, RESOLUTION))
+        label = tf.one_hot(label, depth=NB_CLASSES)
+        return image, label
+
+    return _pp
+
+
+def prepare_dataset(dataset, is_training=True):
+    if is_training:
+        dataset = dataset.shuffle(BATCH_SIZE * 10)
+    dataset = dataset.map(preprocess_dataset(is_training), num_parallel_calls=AUTO)
+    return dataset.batch(BATCH_SIZE).prefetch(AUTO)
+
+
+train_dataset, val_dataset = tfds.load(
+    "tf_flowers", split=["train[:90%]", "train[90%:]"], as_supervised=True
+)
+num_train = train_dataset.cardinality()
+num_val = val_dataset.cardinality()
+print(f"Number of training examples: {num_train}")
+print(f"Number of validation examples: {num_val}")
+
+train_dataset = prepare_dataset(train_dataset, is_training=True)
+val_dataset = prepare_dataset(val_dataset, is_training=False)
+
+"""
+## Implementing the DeiT variants of ViT
+
+Since DeiT is an extension of ViTs it'd make sense to first implement ViT and then extend
+it to support DeiT's components.
+
+First, we'll implement a layer for Stochastic Depth
+([Huang et al.](https://arxiv.org/abs/1603.09382))
+which is used in DeiT for regularization.
+"""
+
+# Referred from: github.com:rwightman/pytorch-image-models.
+class StochasticDepth(layers.Layer):
+    def __init__(self, drop_prop, **kwargs):
+        super().__init__(**kwargs)
+        self.drop_prob = drop_prop
+
+    def call(self, x, training=None):
```

Comment:
Prefer defaulting `training` to a bool, e.g. `True` (it will be passed by the parent anyway...). In the core API we only use the default `None` in order to keep compatibility with the symbolic learning phase in TF1.